src/Core/ProjectFilesWatcher.cs

```diff
@@ -0,0 +1,203 @@
+#region License
+/* 
+ * Copyright (C) 1999-2021 Pavel Tomin.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using Reko.Core.Scripts;
+using Reko.Core.Services;
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Specialized;
+using System.IO;
+using System.Linq;
+using System.Timers;
+
+namespace Reko.Core
+{
+    using FileWatchersMap = ConcurrentDictionary<string, FileSystemWatcher>;
+
+    /// <summary>
+    /// Watch project files, reload them if they were changed.
+    /// </summary>
+    public class ProjectFilesWatcher
+    {
+        private readonly IDecompilerService decompilerSvc;
+        private readonly DecompilerEventListener eventListener;
+        private readonly ILoader loader;
+        private readonly FileWatchersMap scriptWatchers;
+        private readonly ConcurrentDictionary<string, bool> changedFiles;
+        private IDecompiler? decompiler;
+        private Project? project;
+
+        public ProjectFilesWatcher(IServiceProvider services)
+        {
+            this.decompilerSvc = services.RequireService<IDecompilerService>();
+            this.eventListener = services.RequireService<DecompilerEventListener>();
+            this.loader = services.RequireService<ILoader>();
+            this.Decompiler = decompilerSvc.Decompiler;
+            decompilerSvc.DecompilerChanged += OnDecompilerChanged;
+            this.scriptWatchers = new FileWatchersMap();
+            this.changedFiles = new ConcurrentDictionary<string, bool>();
+        }
+
+        public IDecompiler? Decompiler
+        {
+            get => decompiler;
+            set
+            {
+                if (decompiler != null)
+                {
+                    decompiler.ProjectChanged -= OnProjectChanged;
+                }
+                decompiler = value;
+                if (value != null)
+                {
+                    value.ProjectChanged += OnProjectChanged;
+                }
+                this.Project = Decompiler?.Project;
+            }
+        }
+
+        public Project? Project
+        {
+            get => project;
+            set
+            {
+                if (project != null)
+                {
+                    RemoveScripts(project.ScriptFiles.ToList());
```

Comment:
If `RemoveScripts(IList)` is changed to `RemoveScripts(IEnumerable<ScriptFile>)` you can eliminate the 'ToList()` call here.

Reply:
Done in 140746e